File: java/server/src/main/java/com/github/fernthedev/lightchat/server/ClientConnection.java
package com.github.fernthedev.lightchat.server;

import com.github.fernthedev.lightchat.core.StaticHandler;
import com.github.fernthedev.lightchat.core.VersionData;
import com.github.fernthedev.lightchat.core.api.APIUsage;
import com.github.fernthedev.lightchat.core.encryption.UnencryptedPacketWrapper;
import com.github.fernthedev.lightchat.core.encryption.util.EncryptionUtil;
import com.github.fernthedev.lightchat.core.packets.Packet;
import com.github.fernthedev.lightchat.core.packets.latency.LatencyPacket;
import com.github.fernthedev.lightchat.core.packets.latency.PingPacket;
import io.netty.channel.Channel;
import io.netty.channel.ChannelFuture;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NonNull;
import lombok.Setter;
import org.apache.commons.lang3.time.StopWatch;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.lang3.tuple.Pair;

import javax.crypto.Cipher;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.SecretKey;
import java.net.InetSocketAddress;
import java.security.KeyPair;
import java.security.NoSuchAlgorithmException;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

/**
 * Handles client data
 */
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class ClientConnection implements SenderInterface, AutoCloseable {

    private final Server server;

    @Getter
    private boolean connected;

    @Getter
    @Setter
    private boolean registered = false;

    @Getter
    private final EventListener eventListener;

    @Getter
    private String os;

    @Getter
    private String langFramework;

    @EqualsAndHashCode.Include()
    @Getter
    private final UUID uuid;

    @Getter
    @Setter
    private VersionData versionData;



    @Getter
    private final Channel channel;

    @Setter
    private String deviceName;

    /**
     * The keypair encryption
     * Used in initial connection establishment
     *
     * Used before secret key is initialized
     */
    @Getter
    private KeyPair tempKeyPair;


    @Getter
    private SecretKey secretKey;

    @Getter
    private final Cipher encryptCipher;

    @Getter
    private final Cipher decryptCipher;

    /**
     * Packet:[ID,lastPacketSentTime]
     */
    private final Map<Class<? extends Packet>, Pair<Integer, Long>> packetIdMap = new HashMap<>();


    private final StopWatch pingStopWatch = new StopWatch();

    public void setSecretKey(SecretKey secretKey) {
        this.secretKey = secretKey;
        this.tempKeyPair = null;
    }


    public ClientConnection(Server server, Channel channel, UUID uuid, Consumer<ClientConnection> callback) {
        this.server = server;
        this.channel = channel;
        this.uuid = uuid;

        server.getRsaKeyThread().getRandomKey().thenAccept(keyPair -> {
            tempKeyPair = keyPair;
            callback.accept(this);
        });

        eventListener = new EventListener(server, this);

        try {
            encryptCipher = EncryptionUtil.getEncryptCipher();
            decryptCipher = EncryptionUtil.getDecryptCipher();
        } catch (NoSuchPaddingException | NoSuchAlgorithmException e) {
            throw new IllegalStateException(e);
        }
    }


    /**
     *
     * @param timeUnit the time to use
     * @return ping delay of client
     */
    @APIUsage
    public long getPingDelay(TimeUnit timeUnit) {
        return pingStopWatch.getTime(timeUnit);
    }

    private Pair<Integer, Long> updatePacketIdPair(Class<? extends Packet> packet, int newId) {
        Pair<Integer, Long> packetIdPair = packetIdMap.get(packet);

        if (packetIdPair == null)
            packetIdPair = new ImmutablePair<>(0, System.currentTimeMillis());
        else {
            if (newId == -1) newId = packetIdPair.getKey() + 1;

            packetIdPair = new ImmutablePair<>(newId, System.currentTimeMillis());
        }

        packetIdMap.put(packet, packetIdPair);
        return packetIdPair;
    }

    /**
     *
     * @param packet Packet to send
     * @param encrypt if true the packet will be encrypted
     */
    @APIUsage
    public ChannelFuture sendObject(@NonNull Packet packet, boolean encrypt) {
        if (!(packet instanceof LatencyPacket))
            StaticHandler.getCore().getLogger().debug("Sending packet {}:{}", packet.getPacketName(), encrypt);

        Pair<Integer, Long> packetIdPair = updatePacketIdPair(packet.getClass(), -1);

        if (packetIdPair.getLeft() > server.getMaxPacketId() || System.currentTimeMillis() - packetIdPair.getRight() > 900) updatePacketIdPair(packet.getClass(), 0);

        if (encrypt) {
            return channel.writeAndFlush(packet);
        } else {
            return channel.writeAndFlush(new UnencryptedPacketWrapper(packet, packetIdPair.getKey()));
        }
    }



    /**
     * Default usage for {@link #sendObject(Packet, boolean)}
     * @param packet Packet to send
     */
    @APIUsage
    public ChannelFuture sendObject(Packet packet) {
        return sendObject(packet,true);
    }

    /**
     * Closes connection
     */
    public void close() {

        //DISCONNECT FROM SERVER
        server.getLogger().info("Closing player {}", this);

        if (channel != null) {

            channel.close();


            server.getPlayerHandler().getChannelMap().remove(channel);
        }

        connected = false;
        server.getPlayerHandler().getUuidMap().remove(uuid);


        //serverSocket.close();
    }


    @Override
    public String toString() {
        return "[" + getAddress() + "|" + deviceName +"]";
    }


    /**
     *
     * @return the client address
     */
    public String getAddress() {
        if (channel.remoteAddress() == null) {
            return null;
        }

        InetSocketAddress address = (InetSocketAddress) channel.remoteAddress();

        return address.getAddress().toString();
    }

    /**
     * Pings player
     */
    public void ping() {
        pingStopWatch.reset();
        pingStopWatch.start();
        sendObject(new PingPacket(),false);
    }

    @NonNull
    @Override
    public ChannelFuture sendPacket(Packet packet) {
        return sendObject(packet);
    }

    @Override
    public String getName() {
        return deviceName;
    }

    public void finishPing() {
        if (!pingStopWatch.isStopped())
            pingStopWatch.stop();
    }

    /**
     * Packet:[ID,lastPacketSentTime]
     */
    public Pair<Integer, Long> getPacketId(Class<? extends Packet> packet) {
        packetIdMap.computeIfAbsent(packet, aClass -> new ImmutablePair<>(0,(long) -1));

        return packetIdMap.get(packet);
    }


    void finishConstruct(String name, String os, String langFramework) {
        this.deviceName = name;
        this.os = os;
        this.langFramework = langFramework;
    }
}

File: java/server/src/main/java/com/github/fernthedev/lightchat/server/ClientConnection.kt
package com.github.fernthedev.lightchat.server

import com.github.fernthedev.lightchat.core.VersionData
import com.github.fernthedev.lightchat.core.api.APIUsage
import com.github.fernthedev.lightchat.core.codecs.AcceptablePacketTypes
import com.github.fernthedev.lightchat.core.encryption.*
import com.github.fernthedev.lightchat.core.encryption.util.EncryptionUtil
import com.github.fernthedev.lightchat.core.packets.latency.PingPacket
import io.netty.channel.Channel
import kotlinx.coroutines.*
import org.apache.commons.lang3.time.StopWatch
import java.net.InetSocketAddress
import java.security.KeyPair
import java.security.SecureRandom
import java.util.*
import java.util.concurrent.TimeUnit
import javax.crypto.Cipher
import javax.crypto.SecretKey

/**
 * Handles client data
 */
@OptIn(ExperimentalCoroutinesApi::class)
class ClientConnection(
    private val server: Server,
    val channel: Channel,
    val uuid: UUID
) : SenderInterface, AutoCloseable {
    var connected = false
        private set

    var registered = false

    val eventListener: EventListener = EventListener(server, this)

    lateinit var os: String
        private set

    lateinit var langFramework: String
        private set


    lateinit var versionData: VersionData


    override lateinit var name: String
        private set

    /**
     * The keypair encryption
     * Used in initial connection establishment
     *
     * Used before secret key is initialized
     */
    var tempKeyPair: KeyPair? = null
        private set

    var secretKey: SecretKey? = null
        set(value) {
            requireNotNull(value)
            field = value
            tempKeyPair = null
            secureRandom = EncryptionUtil.getSecureRandom(value)
        }

    var encryptCipher: ThreadLocal<Cipher>
        private set


    var decryptCipher: ThreadLocal<Cipher>
        private set

    var secureRandom: SecureRandom? = null
        private set

    fun encryptionRegistered(): Boolean {
        return secretKey != null
    }

    /**
     * Packet:[ID,lastPacketSentTime]
     */
    private val packetJSONIdMap: MutableMap<Class<out AcceptablePacketTypes>, Pair<Int, Long>> = HashMap()
    private val pingStopWatch = StopWatch()


    init {
        encryptCipher = ThreadLocal.withInitial { EncryptionUtil.generateEncryptCipher() }
        decryptCipher = ThreadLocal.withInitial { EncryptionUtil.generateDecryptCipher() }
    }

    /**
     *
     * @param timeUnit the time to use
     * @return ping delay of client
     */
    @APIUsage
    fun getPingDelay(timeUnit: TimeUnit?): Long {
        return pingStopWatch.getTime(timeUnit)
    }

    private fun updatePacketIdPair(packetJSON: Class<out AcceptablePacketTypes>, newId: Int): Pair<Int, Long> {
        var copiedId = newId
        var packetIdPair = packetJSONIdMap[packetJSON]
        if (packetIdPair == null) {
            packetIdPair = Pair(0, System.currentTimeMillis())
        } else {
            if (copiedId == -1) {
                copiedId = packetIdPair.first + 1
            }
            packetIdPair = Pair(copiedId, System.currentTimeMillis())
        }
        packetJSONIdMap[packetJSON] = packetIdPair
        return packetIdPair
    }


    /**
     * Closes connection
     */
    override fun close() {

        //DISCONNECT FROM SERVER
        server.logger.info("Closing player {}", this)
        channel.close()
        server.playerHandler.channelMap.remove(channel)
        connected = false
        server.playerHandler.uuidMap.remove(uuid)
    }

    override fun toString(): String {
        return "[$address|$name]"
    }

    val address: String?
        /**
         *
         * @return the client address
         */
        get() {
            if (channel.remoteAddress() == null) {
                return null
            }
            val address = channel.remoteAddress() as InetSocketAddress
            return address.address.toString()
        }

    /**
     * Pings player
     */
    suspend fun ping() {
        pingStopWatch.reset()
        pingStopWatch.start()
        sendPacketLaunch(PingPacket().transport(false))
    }

    override suspend fun sendPacketDeferred(transporter: PacketTransporter) = coroutineScope {
        val packet = transporter.packet
        val packetIdPair = updatePacketIdPair(packet.javaClass, -1)
        if (packetIdPair.first > server.maxPacketId || System.currentTimeMillis() - packetIdPair.second > 900) {
            updatePacketIdPair(
                packet.javaClass,
                0
            )
        }
        return@coroutineScope async(Dispatchers.IO) {
            return@async channel.writeAndFlush(transporter)
        }
    }

    override suspend fun sendPacketLaunch(transporter: PacketTransporter): Job = coroutineScope {
        val packet = transporter.packet
        val packetIdPair = updatePacketIdPair(packet.javaClass, -1)
        if (packetIdPair.first > server.maxPacketId || System.currentTimeMillis() - packetIdPair.second > 900) {
            updatePacketIdPair(
                packet.javaClass,
                0
            )
        }
        return@coroutineScope launch(Dispatchers.IO) {
            channel.writeAndFlush(transporter)
        }
    }

    fun finishPing() {
        if (!pingStopWatch.isStopped) pingStopWatch.stop()
    }

    /**
     * Packet:[ID,lastPacketSentTime]
     */
    fun getPacketId(packetJSON: Class<out AcceptablePacketTypes>): Pair<Int, Long> {
        packetJSONIdMap.computeIfAbsent(packetJSON) { Pair(0, -1L) }
        return packetJSONIdMap[packetJSON]!!
    }

    fun finishConstruct(name: String, os: String, langFramework: String) {
        this.name = name
        this.os = os
        this.langFramework = langFramework
    }

    internal fun setupKeypair(key: KeyPair) {
        tempKeyPair = key
    }
}